source/azure_iot_hub_client.c

```diff
@@ -0,0 +1,538 @@
+    }
+
+    if( ( xResult = AzureIoTMQTT_Disconnect( &( xAzureIoTHubClientHandle->_internal.xMQTTContext ) ) ) != AzureIoTMQTTSuccess )
+    {
+        AZLogError( ( "Failed to disconnect: %d\r\n", xResult ) );
+        ret = eAzureIoTHubClientFailed;
+    }
+    else
+    {
+        AZLogInfo( ( "Disconnecting the MQTT connection with %.*s.\r\n", xAzureIoTHubClientHandle->_internal.ulHostnameLength,
+                     ( const char * ) xAzureIoTHubClientHandle->_internal.pucHostname ) );
+        ret = eAzureIoTHubClientSuccess;
+    }
+
+    return ret;
+}
+/*-----------------------------------------------------------*/
+
+AzureIoTHubClientResult_t AzureIoTHubClient_SendTelemetry( AzureIoTHubClientHandle_t xAzureIoTHubClientHandle,
+                                                           const uint8_t * pucTelemetryData,
+                                                           uint32_t ulTelemetryDataLength,
+                                                           AzureIoTMessageProperties_t * pxProperties )
+{
+    AzureIoTMQTTStatus_t xResult;
+    AzureIoTHubClientResult_t ret;
+    AzureIoTMQTTPublishInfo_t xMQTTPublishInfo = { 0 };
+    uint16_t usPublishPacketIdentifier;
+    size_t telemetry_topic_length;
+    az_result res;
+
+    if( xAzureIoTHubClientHandle == NULL )
+    {
+        AZLogError( ( "Failed to send telemetry: Invalid argument \r\n" ) );
+        ret = eAzureIoTHubClientInvalidArgument;
+    }
+    else if( az_result_failed( res = az_iot_hub_client_telemetry_get_publish_topic( &xAzureIoTHubClientHandle->_internal.xAzureIoTHubClientCore,
+                                                                                    ( pxProperties != NULL ) ? &pxProperties->_internal.properties : NULL,
+                                                                                    ( char * ) xAzureIoTHubClientHandle->_internal.pucAzureIoTHubClientWorkingBuffer,
+                                                                                    xAzureIoTHubClientHandle->_internal.ulAzureIoTHubClientWorkingBufferLength,
+                                                                                    &telemetry_topic_length ) ) )
+    {
+        AZLogError( ( "Failed to get telemetry topic : %08x \r\n", res ) );
+        ret = eAzureIoTHubClientFailed;
+    }
+    else
+    {
+        xMQTTPublishInfo.qos = AzureIoTMQTTQoS1;
+        xMQTTPublishInfo.pTopicName = ( const char * ) xAzureIoTHubClientHandle->_internal.pucAzureIoTHubClientWorkingBuffer;
+        xMQTTPublishInfo.topicNameLength = ( uint16_t ) telemetry_topic_length;
+        xMQTTPublishInfo.pPayload = ( const void * ) pucTelemetryData;
+        xMQTTPublishInfo.payloadLength = ulTelemetryDataLength;
+
+        /* Get a unique packet id. */
+        usPublishPacketIdentifier = AzureIoTMQTT_GetPacketId( &( xAzureIoTHubClientHandle->_internal.xMQTTContext ) );
+
+        /* Send PUBLISH packet. Packet ID is not used for a QoS1 publish. */
```

Comment:
I don't understand comment about PackeID not used, but we use it on the next line.  I thought it was needed for PUBACK?

Reply:
Yea I think this is an erroneous comment. @hihigupt to confirm.

https://github.com/FreeRTOS/coreMQTT/blob/f756d9833ad79b60084def59935fdabf475c08b0/source/include/core_mqtt.h#L501

Reply:
Wrong comment

---
In reply to: [608951348](https://github.com/Azure/azure-iot-middleware-freertos/pull/60#discussion_r608951348) [](ancestors = 608951348)